### PR TITLE
Grammar from type declaration

### PR DESCRIPTION
**Pull request by pdasigi**
I mainly added a method `type_declaration.get_valid_actions()` that takes a name mapping, type signatures and a set of basic types and produces a dict from all non-terminals to productions starting from them.

@matt-gardner I tested this on the NLVR type declaration, and it works, but with wikitables, it has some problems with placeholder types. I also need to test it with table specific mappings.  I'll fix those soon. But I'm making this PR now to show you the API, so that you can continue with the stuff you're doing.

### CODE REVIEW

**Comment by pdasigi** [comment]
This should now be ready for review.

File: allennlp/data/semparse/type_declarations/type_declaration.py [hunk 1]
```diff
@@ -254,3 +257,128 @@ def make_VariableExpression(self, name):
...
+    return "%s -> %s" % (str(source), str(target))
+
+
+def _get_complex_type_productions(complex_type: ComplexType) -> List[Tuple[str, str]]:
+    """
+    Takes a complex type without any placeholders and returns all productions that lead to it, starting
+    from the most basic return type. For example, if the complex is `<a,<<b,c>,d>>`, this gives the
+    following tuples
+    ('<<b,c>,d>', '<<b,c>,d> -> [<a,<<b,c>,d>>, a]')
+    ('d', 'd -> [<<b,c>,d>, <b,c>]')
```
Comment (matt-gardner) [inline allennlp/data/semparse/type_declarations/type_declaration.py]: Why doesn't this also give `('c', 'c -> [<b,c>, b]')`?  Or does it?  It looks to me like the code will produce this; maybe you just missed it in the docstring?

File: allennlp/data/semparse/type_declarations/type_declaration.py [hunk 2]
```diff
@@ -254,3 +257,128 @@ def make_VariableExpression(self, name):
...
+        complex_type = complex_type.second
+    return all_productions
+
+
+def get_valid_actions(name_mapping: Dict[str, str],
+                      type_signatures: Dict[str, Type],
+                      basic_types: Set[Type]) -> Dict[str, Set[str]]:
+    """
+    Generates all the valid actions starting from each non-terminal. For terminals of a specific type, we
+    simply add to valid actions, productions from thee types to the terminals. Among those types, we keep
```
Comment (matt-gardner) [inline allennlp/data/semparse/type_declarations/type_declaration.py]: `s/thee/the/`?

File: allennlp/data/semparse/type_declarations/type_declaration.py [hunk 3]
```diff
@@ -254,3 +257,128 @@ def make_VariableExpression(self, name):
...
+    track of all the non-basic types (i.e., function types). For those types, we infer the list of productions
+    that start from a basic type leading to them.
+    For complex types that do not contain ANY_TYPE or placeholder
```
Comment (matt-gardner) [inline allennlp/data/semparse/type_declarations/type_declaration.py]: Do you use vim?  I can't remember if you do or not.  If you do, can you set your textwidth to 99, then run `gwap` on this paragraph?  You'd need to add blank lines above and below the paragraph, so that `gwap` works right, then remove them.

Anyway, the point is that the docstring is a little too long on some lines, and too short on others.  Vim will easily fix this for you; not sure how to do it with other editors.

File: allennlp/data/semparse/type_declarations/type_declaration.py [hunk 4]
```diff
@@ -254,3 +257,128 @@ def make_VariableExpression(self, name):
...
+    types, this is straight-forward. For example, if the complex type is <e,<r,<d,r>>>, the productions should
+    be [r -> [<d,r>, r], <d,r> -> [<r,<d,r>>, r], <r,<d,r>> -> [<e,<r,<d,r>>>, e]].
+    However, if the complex type contains ANY_TYPE, we substitute it with all possible basic types. If it
+    contains placeholders, the substituyions need to be constrained. For example, for <#1,#1>, <e,r> is not a
```
Comment (matt-gardner) [inline allennlp/data/semparse/type_declarations/type_declaration.py]: `s/substituyions/substitutions/`

File: allennlp/data/semparse/type_declarations/type_declaration.py [hunk 5]
```diff
@@ -254,3 +257,128 @@ def make_VariableExpression(self, name):
...
+    valid substitution.
+    """
+    valid_actions: Dict[str, Set[str]] = defaultdict(set)
+
+    complex_types = set()
+    for name, alias in name_mapping.items():
+        if not alias in type_signatures:
```
Comment (matt-gardner) [inline allennlp/data/semparse/type_declarations/type_declaration.py]: It's more idiomatic to do `if alias not in type_signatures:`

File: allennlp/data/semparse/type_declarations/type_declaration.py [hunk 6]
```diff
@@ -254,3 +257,128 @@ def make_VariableExpression(self, name):
...
+    simply add to valid actions, productions from thee types to the terminals. Among those types, we keep
+    track of all the non-basic types (i.e., function types). For those types, we infer the list of productions
+    that start from a basic type leading to them.
+    For complex types that do not contain ANY_TYPE or placeholder
+    types, this is straight-forward. For example, if the complex type is <e,<r,<d,r>>>, the productions should
+    be [r -> [<d,r>, r], <d,r> -> [<r,<d,r>>, r], <r,<d,r>> -> [<e,<r,<d,r>>>, e]].
+    However, if the complex type contains ANY_TYPE, we substitute it with all possible basic types. If it
+    contains placeholders, the substituyions need to be constrained. For example, for <#1,#1>, <e,r> is not a
+    valid substitution.
+    """
```
Comment (matt-gardner) [inline allennlp/data/semparse/type_declarations/type_declaration.py]: Can you add a `Parameters` section here describing what your inputs are?

File: allennlp/data/semparse/type_declarations/type_declaration.py (hunk 5 above)
Comment (matt-gardner) [reply allennlp/data/semparse/type_declarations/type_declaration.py]: Also, when would this happen, where you have a name mapping that doesn't show up in `type_signatures`?

File: allennlp/data/semparse/type_declarations/type_declaration.py [hunk 7]
```diff
@@ -254,3 +257,128 @@ def make_VariableExpression(self, name):
...
+        name_type = type_signatures[alias]
+        # Type to terminal productions.
+        for substituted_type in _substitute_any_type(name_type, basic_types):
+            valid_actions[str(substituted_type)].add(_make_production_string(substituted_type, name))
+        # Keeping track of complex types.
+        if isinstance(name_type, ComplexType) and name_type != ANY_TYPE:
+            complex_types.add(name_type)
+
+    for complex_type in complex_types:
+        if isinstance(complex_type, PlaceholderType):
```
Comment (matt-gardner) [inline allennlp/data/semparse/type_declarations/type_declaration.py]: This is quite a lot of nesting; it's hard to follow.  I'd recommend splitting this out into another function, like `_add_placeholder_actions(complex_type, valid_actions)`.

File: allennlp/data/semparse/type_declarations/type_declaration.py [hunk 8]
```diff
@@ -254,3 +257,128 @@ def make_VariableExpression(self, name):
...
+                    for basic_type in basic_types:
+                        application_type = complex_type.get_application_type(basic_type)
+                        valid_actions[str(application_type)].add(_make_production_string(application_type,
+                                                                                         [complex_type,
+                                                                                          basic_type]))
+                        for head, production in _get_complex_type_productions(application_type):
+                            valid_actions[head].add(production)
+                else:
+                    # This means complex_type.first is ComplexType(ANY_TYPE, ANY_TYPE)
+                    # TODO(pradeep): Assuming this is a reverse type. That is the only type where the
```
Comment (matt-gardner) [inline allennlp/data/semparse/type_declarations/type_declaration.py]: Instead of a `TODO`, can you check that `complex_type` actually _is_ a reverse type, and raise a `NotImplementedError` otherwise?  Better to crash than to fail silently in cases where you're explicitly not supporting something.

File: allennlp/data/semparse/type_declarations/type_declaration.py [hunk 9]
```diff
@@ -254,3 +257,128 @@ def make_VariableExpression(self, name):
...
+            substitutions.add(ComplexType(first_type, second_type))
+    return substitutions
+
+
+def _substitute_placeholder_type(_type: Type, basic_type: BasicType) -> Type:
+    """
+    Takes a type with placeholders, and a basic type and substitutes all occurrences of the placeholder with
+    that type.
+    """
+    # TODO (pradeep): This assumes there's just one placeholder in the type. So this doesn't work with
```
Comment (matt-gardner) [inline allennlp/data/semparse/type_declarations/type_declaration.py]: Is there a way to make this crash if there is more than one placeholder?

File: allennlp/data/semparse/type_declarations/type_declaration.py [hunk 10]
```diff
@@ -254,3 +257,128 @@ def make_VariableExpression(self, name):
...
+                        for head, production in _get_complex_type_productions(application_type):
+                            valid_actions[head].add(production)
+                else:
+                    # This means complex_type.first is ComplexType(ANY_TYPE, ANY_TYPE)
+                    # TODO(pradeep): Assuming this is a reverse type. That is the only type where the
+                    # input type is a ComplexType for now. But this needs to be more general later.
+                    for first_type in basic_types:
+                        for second_type in basic_types:
+                            input_type = ComplexType(first_type, second_type)
+                            application_type = complex_type.get_application_type(input_type)
```
Comment (matt-gardner) [inline allennlp/data/semparse/type_declarations/type_declaration.py]: What does `get_application_type` do?

File: tests/data/semparse/type_declarations/type_declaration_test.py [hunk 11]
```diff
@@ -8,3 +11,57 @@ def test_basic_types_conflict_on_names(self):
         type_a = types.NamedBasicType("A")
         type_b = types.NamedBasicType("B")
         assert type_a.resolve(type_b) is None
+
+    def test_valid_actions(self):
```
Comment (matt-gardner) [inline tests/data/semparse/type_declarations/type_declaration_test.py]: Can you change all of these from `test_valid_actions` to `test_get_valid_actions`?

File: tests/data/semparse/type_declarations/type_declaration_test.py [hunk 12]
```diff
@@ -8,3 +11,57 @@ def test_basic_types_conflict_on_names(self):
...
+        type_e = types.NamedBasicType("E")
+        name_mapping = {'sample_function': 'F'}
+        # <e,<r,<d,r>>>
+        type_signatures = {'F': ComplexType(type_e, ComplexType(type_r, ComplexType(type_d, type_r)))}
+        basic_types = {type_r, type_d, type_e}
+        valid_actions = types.get_valid_actions(name_mapping, type_signatures, basic_types)
+        assert len(valid_actions) == 4
+        assert valid_actions["<e,<r,<d,r>>>"] == {"<e,<r,<d,r>>> -> sample_function"}
+        assert valid_actions["<r,<d,r>>"] == {"<r,<d,r>> -> [<e,<r,<d,r>>>, e]"}
+        assert valid_actions["<d,r>"] == {"<d,r> -> [<r,<d,r>>, r]"}
```
Comment (matt-gardner) [inline tests/data/semparse/type_declarations/type_declaration_test.py]: Hardy har har...  https://www.youtube.com/watch?v=HcD1oA_s4-U - your type is `[<r,<d,r>>, r]`.  It's like you did that on purpose... =)

File: tests/data/semparse/type_declarations/type_declaration_test.py [hunk 13]
```diff
@@ -8,3 +11,57 @@ def test_basic_types_conflict_on_names(self):
...
+        type_r = types.NamedBasicType("R")
+        type_d = types.NamedBasicType("D")
+        type_e = types.NamedBasicType("E")
+        name_mapping = {'sample_function': 'F'}
+        # <#1,#1>
+        type_signatures = {'F': types.IdentityType(types.ANY_TYPE, types.ANY_TYPE)}
+        basic_types = {type_r, type_d, type_e}
+        valid_actions = types.get_valid_actions(name_mapping, type_signatures, basic_types)
+        assert len(valid_actions) == 4
+        assert valid_actions["<#1,#1>"] == {"<#1,#1> -> sample_function"}
```
Comment (matt-gardner) [inline tests/data/semparse/type_declarations/type_declaration_test.py]: Does it make sense to add rules like `<e,e> -> sample_function`, and `e -> [<e,e>, e]`?  Or will leaving the placeholders in always be sufficient?

File: tests/data/semparse/type_declarations/type_declaration_test.py [hunk 14]
```diff
@@ -8,3 +11,57 @@ def test_basic_types_conflict_on_names(self):
...
+        type_r = types.NamedBasicType("R")
+        type_d = types.NamedBasicType("D")
+        type_e = types.NamedBasicType("E")
+        name_mapping = {'sample_function': 'F'}
+        # <#1,r>
+        type_signatures = {'F': ComplexType(types.ANY_TYPE, type_r)}
+        basic_types = {type_r, type_d, type_e}
+        valid_actions = types.get_valid_actions(name_mapping, type_signatures, basic_types)
+        assert len(valid_actions) == 4
+        assert valid_actions["<d,r>"] == {"<d,r> -> sample_function"}
```
Comment (matt-gardner) [inline tests/data/semparse/type_declarations/type_declaration_test.py]: Related to the above, why are these instantiated to basic types while the placeholders aren't?  Why not just leave it as `<ANY,r> -> sample_function`?

File: allennlp/data/semparse/type_declarations/type_declaration.py [hunk 15]
```diff
@@ -254,3 +257,128 @@ def make_VariableExpression(self, name):
             else:
                 raise RuntimeError("Unknown prefix: %s. Did you forget to pass it to the constructor?" % prefix)
         return super(DynamicTypeLogicParser, self).make_VariableExpression(name)
+
+
+def _substitute_any_type(_type: Type, basic_types: Set[BasicType]) -> Set[Type]:
+    """
+    Takes a type, and a set of basic types and substitutes all instances of ANY_TYPE with all possible basic
```
[the remark this answers is not in the document]
Comment (pdasigi) [reply allennlp/data/semparse/type_declarations/type_declaration.py]: Done.

File: allennlp/data/semparse/type_declarations/type_declaration.py [hunk 16]
```diff
@@ -254,3 +257,128 @@ def make_VariableExpression(self, name):
...
+    substitutions = set()
+    for first_type in _substitute_any_type(_type.first, basic_types):
+        for second_type in _substitute_any_type(_type.second, basic_types):
+            substitutions.add(ComplexType(first_type, second_type))
+    return substitutions
+
+
+def _substitute_placeholder_type(_type: Type, basic_type: BasicType) -> Type:
+    """
+    Takes a type with placeholders, and a basic type and substitutes all occurrences of the placeholder with
```
[the remark this answers is not in the document]
Comment (pdasigi) [reply allennlp/data/semparse/type_declarations/type_declaration.py]: Done.

File: allennlp/data/semparse/type_declarations/type_declaration.py (hunk 9 above)
Comment (pdasigi) [reply allennlp/data/semparse/type_declarations/type_declaration.py]: Done.

File: allennlp/data/semparse/type_declarations/type_declaration.py (hunk 1 above)
Comment (pdasigi) [reply allennlp/data/semparse/type_declarations/type_declaration.py]: It didn't do that earlier. I expanded only the branches that lead to the complex type, but I could have expanded the others as well. Changed this now. Thanks for pointing this out!

File: allennlp/data/semparse/type_declarations/type_declaration.py (hunk 2 above)
Comment (pdasigi) [reply allennlp/data/semparse/type_declarations/type_declaration.py]: Done.

File: allennlp/data/semparse/type_declarations/type_declaration.py (hunk 3 above)
Comment (pdasigi) [reply allennlp/data/semparse/type_declarations/type_declaration.py]: Done. I use vim.

File: allennlp/data/semparse/type_declarations/type_declaration.py (hunk 4 above)
Comment (pdasigi) [reply allennlp/data/semparse/type_declarations/type_declaration.py]: Done.

File: allennlp/data/semparse/type_declarations/type_declaration.py (hunk 6 above)
Comment (pdasigi) [reply allennlp/data/semparse/type_declarations/type_declaration.py]: Done.

File: allennlp/data/semparse/type_declarations/type_declaration.py (hunk 5 above)
Comment (pdasigi) [reply allennlp/data/semparse/type_declarations/type_declaration.py]: This happens only when the `name` is `lambda`. Changed the conditional to simply check for that.

File: allennlp/data/semparse/type_declarations/type_declaration.py (hunk 7 above)
Comment (pdasigi) [reply allennlp/data/semparse/type_declarations/type_declaration.py]: Done.

File: allennlp/data/semparse/type_declarations/type_declaration.py (hunk 10 above)
Comment (pdasigi) [reply allennlp/data/semparse/type_declarations/type_declaration.py]: It gets the return type when the complex type is applied on a given argument type. Added a comment.

File: allennlp/data/semparse/type_declarations/type_declaration.py (hunk 8 above)
Comment (pdasigi) [reply allennlp/data/semparse/type_declarations/type_declaration.py]: Done.

File: tests/data/semparse/type_declarations/type_declaration_test.py (hunk 11 above)
Comment (pdasigi) [reply tests/data/semparse/type_declarations/type_declaration_test.py]: Done.

File: tests/data/semparse/type_declarations/type_declaration_test.py (hunk 12 above)
Comment (pdasigi) [reply tests/data/semparse/type_declarations/type_declaration_test.py]: :smile:

File: tests/data/semparse/type_declarations/type_declaration_test.py (hunk 14 above)
Comment (pdasigi) [reply tests/data/semparse/type_declarations/type_declaration_test.py]: I do not foresee defining functions with `ANY_TYPE` like I do in this test. We should just use `PlaceholderType` instead. The purpose of this test is to check whether `ANY_TYPE` get substituted by every possible type, and the idea is to simulate an intermediate step while building up a tree to get the actions for a `PlaceholderType`.

In general, I think of `ANY_TYPE`s as those that get resolved during parsing, and placeholders as those that do not. This means that we'll have placeholder types only within terminal productions.

Comment (pdasigi) [reply tests/data/semparse/type_declarations/type_declaration_test.py]: Added a comment within this test.

File: tests/data/semparse/type_declarations/type_declaration_test.py (hunk 13 above)
Comment (pdasigi) [reply tests/data/semparse/type_declarations/type_declaration_test.py]: Practically, leaving them as placeholders should always be sufficient. The action sequences contain placeholder strings in terminal productions, and they appear as resolved types after they are applied to other types: https://github.com/allenai/allennlp/blob/wikitables/tests/data/semparse/worlds/wikitables_world_test.py#L32-L37

File: allennlp/data/semparse/type_declarations/type_declaration.py [hunk 17]
```diff
@@ -310,30 +313,82 @@ def _get_complex_type_productions(complex_type: ComplexType) -> List[Tuple[str,
         all_productions.append((str(complex_type.second), _make_production_string(complex_type.second,
                                                                                   [complex_type,
                                                                                    complex_type.first])))
+        for production in _get_complex_type_productions(complex_type.first):
+            all_productions.append(production)
         complex_type = complex_type.second
     return all_productions
 
 
+def _get_placeholder_actions(complex_type, basic_types, valid_actions) -> None:
```
Comment (matt-gardner) [inline allennlp/data/semparse/type_declarations/type_declaration.py]: Type annotations on the arguments here.

File: allennlp/data/semparse/type_declarations/type_declaration.py [hunk 18]
```diff
@@ -310,30 +313,82 @@ def _get_complex_type_productions(complex_type: ComplexType) -> List[Tuple[str,
...
+    placeholder substitution.
+
+    Parameters
+    ----------
+    name_mapping : ``Dict[str, str]``
+        The mapping of names that appear in your logical form languages to their aliases for NLTK.
+        If you are getting all valid actions for a type declaration, this can be the
+        ``COMMON_NAME_MAPPING``.
+    type_signatures : ``Dict[str, Type]``
+        The ampping from name aliases to their types. If you are getting all valid actions for a
```
Comment (matt-gardner) [inline allennlp/data/semparse/type_declarations/type_declaration.py]: `s/ampping/mapping/`
